File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.py
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node: return 0

            nonlocal result
            lCoin = dfs(node.left)
            rCoin = dfs(node.right)
            coins = node.val - 1 + lCoin + rCoin
            result += abs(coins)
            return coins


        result = 0
        dfs(root)
        return result

File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.java
class Solution {
    private int result;

    private int dfs(TreeNode node) {
        if (node == null) return 0;

        int lCoin = dfs(node.left);
        int rCoin = dfs(node.right);
        int coins = node.val - 1 + lCoin + rCoin;
        result += Math.abs(coins);
        return coins;
    }

    public int distributeCoins(TreeNode root) {
        result = 0;
        dfs(root);
        return result;
    }
}